File: java-advent-of-code/src/main/java/com/mycompany/advent_of_code_java/Advent_of_code_2025_001_002.java
package com.mycompany.advent_of_code_java;

import module java.base;

public class Advent_of_code_2025_001_002 {
    private static final int MOD = 100;

    public static void main(String[] args) {
        String inputPath = (args.length >= 1)
                ? args[0]
                : "E:\\documents-container\\advent-of-code\\java-advent-of-code\\inputs\\2025-001\\input.txt";

        try {
            long password = solve(Path.of(inputPath));
            System.out.println(password);
        } catch (IOException e) {
            System.err.println("Failed to read input: " + inputPath);
            e.printStackTrace(System.err);
            System.exit(1);
        } catch (IllegalArgumentException e) {
            System.err.println("Invalid input format: " + e.getMessage());
            System.exit(2);
        }
    }

    public static long solve(Path inputFile) throws IOException {
        int pos = 50;
        long zeroClicks = 0;

        try (BufferedReader br = Files.newBufferedReader(inputFile)) {
            String line;
            long lineNo = 0;

            while ((line = br.readLine()) != null) {
                lineNo++;
                line = line.trim();
                if (line.isEmpty()) continue;

                char dir = line.charAt(0);
                if (dir != 'L' && dir != 'R') {
                    throw new IllegalArgumentException("Line " + lineNo + ": must start with L or R, got: " + line);
                }

                String distStr = line.substring(1).trim();
                if (distStr.isEmpty()) {
                    throw new IllegalArgumentException("Line " + lineNo + ": missing distance: " + line);
                }

                long dist;
                try {
                    dist = Long.parseLong(distStr);
                } catch (NumberFormatException nfe) {
                    throw new IllegalArgumentException("Line " + lineNo + ": bad distance: " + line);
                }
                if (dist < 0) {
                    throw new IllegalArgumentException("Line " + lineNo + ": distance must be non-negative: " + line);
                }
                if (dist == 0) continue; // no clicks, no movement

                // Count how many clicks during this rotation land on 0
                zeroClicks += countHitsZeroDuringRotation(pos, dir, dist);

                // Update final position after full rotation
                long distMod = dist % MOD;
                if (dir == 'R') {
                    pos = (int) ((pos + distMod) % MOD);
                } else { // 'L'
                    pos = (int) ((pos - distMod) % MOD);
                    if (pos < 0) pos += MOD;
                }
            }
        }

        return zeroClicks;
    }

    /**
     * Counts the number of click-steps k in [1..dist] such that the dial points at 0
     * immediately after that click, starting from 'pos' before the rotation.
     */
    private static long countHitsZeroDuringRotation(int pos, char dir, long dist) {
        // Find the first k (1..100) that hits 0, then every 100 after that.
        int kFirst;

        if (dir == 'R') {
            // pos + k ≡ 0 (mod 100)  =>  k ≡ -pos ≡ (100 - pos) (mod 100)
            int k0 = (MOD - (pos % MOD)) % MOD; // in [0..99]
            kFirst = (k0 == 0) ? MOD : k0;      // if 0, next hit is at 100
        } else {
            // pos - k ≡ 0 (mod 100)  =>  k ≡ pos (mod 100)
            int k0 = pos % MOD;                // in [0..99]
            kFirst = (k0 == 0) ? MOD : k0;      // if 0, next hit is at 100
        }

        if (kFirst > dist) return 0;

        // hits at kFirst, kFirst+100, kFirst+200, ... <= dist
        return 1 + (dist - kFirst) / MOD;
    }
}

File: python/Advent_of_code_2025_001_002.py
import sys
from pathlib import Path

def count_hits_zero_during_rotation(pos: int, direction: str, dist: int) -> int:
    """
    Counts the number of click-steps k in [1..dist] such that the dial points at 0
    immediately after that click, starting from 'pos' before the rotation.
    """
    MOD = 100
    
    if direction == 'R':
        # pos + k ≡ 0 (mod 100)  =>  k ≡ -pos ≡ (100 - pos) (mod 100)
        k0 = (MOD - (pos % MOD)) % MOD
        k_first = MOD if k0 == 0 else k0
    else: # 'L'
        # pos - k ≡ 0 (mod 100)  =>  k ≡ pos (mod 100)
        k0 = pos % MOD
        k_first = MOD if k0 == 0 else k0

    if k_first > dist:
        return 0

    # hits at k_first, k_first+100, k_first+200, ... <= dist
    return 1 + (dist - k_first) // MOD

def solve(input_file_path: Path) -> int:
    MOD = 100
    pos = 50
    zero_clicks = 0
    
    with open(input_file_path, 'r', encoding='utf-8') as f:
        for line_no, line in enumerate(f, 1):
            line = line.strip()
            if not line:
                continue
                
            dir_char = line[0]
            if dir_char not in ('L', 'R'):
                raise ValueError(f"Line {line_no}: must start with L or R, got: {line}")
            
            dist_str = line[1:].strip()
            if not dist_str:
                raise ValueError(f"Line {line_no}: missing distance: {line}")
            
            try:
                dist = int(dist_str)
            except ValueError:
                raise ValueError(f"Line {line_no}: bad distance: {line}")
                
            if dist < 0:
                raise ValueError(f"Line {line_no}: distance must be non-negative: {line}")
            if dist == 0:
                continue

            # Count how many clicks during this rotation land on 0
            zero_clicks += count_hits_zero_during_rotation(pos, dir_char, dist)

            # Update final position after full rotation
            dist_mod = dist % MOD
            if dir_char == 'R':
                pos = (pos + dist_mod) % MOD
            else: # 'L'
                pos = (pos - dist_mod) % MOD
                # Python's % handles negative numbers, but we'll be explicit to match Java
                if pos < 0:
                    pos += MOD
                    
    return zero_clicks

def main():
    # Default path based on your environment
    default_path = r"E:\documents-container\advent-of-code\java-advent-of-code\inputs\2025-001\input.txt"
    
    input_path_str = sys.argv[1] if len(sys.argv) >= 2 else default_path
    input_path = Path(input_path_str)
    
    try:
        password = solve(input_path)
        print(password)
    except FileNotFoundError:
        print(f"Failed to read input: {input_path}", file=sys.stderr)
        sys.exit(1)
    except ValueError as e:
        print(f"Invalid input format: {e}", file=sys.stderr)
        sys.exit(2)

if __name__ == "__main__":
    main()